Introduce `secrets.properties`. Use `gradle.properties` only for build configuration.

wzieba opened the pull request:
### Description

This PR introduces a split of `gradle.properties` into two files: `secrets.properties` to contain only sensitive data and `gradle.properties` to contain only build configuration. More details internally: pbArwn-6R4-p2

-----

`dependency-tree-diff` step won't succeed because we changed the build steps comparing to `trunk` (removed copying `gradle.properties-example`)

## To Test:

### Signing configuration

1. Checkout `trunk` and setup all secrets `bundle exec fastlane run configure_apply`. 
2. Run `:WordPress:signingReport` task, copy the content
3. Checkout this branch, run `:WordPress:signingReport` again
4. Compare outputs of both executions. They should be the same.

### App secrets

Open generated `BuildConfig.kt` file. Make sure it contains valid secrets. You can compare content of `BuildConfig.kt` from build from `trunk`, but be aware they'll be different: unused secrets have been removed from mobile secrets.

### Measure builds

Apply secrets `bundle exec fastlane run configure_apply` and assert, that at the end of the task ` Build time report of Xm Ys has been received by Apps Metrics.` is printed to the console.

wzieba committed:
Remove unused build config fields

wzieba committed:
Introduce defaults.properties. Load defaults and secrets when building the app

wzieba committed:
Sort defaults.properties alphabetically

wzieba committed:
Introduce secrets.properties from mobile secrets

wzieba committed:
Introduce `gradle.properties` file with only build configuration data. Remove example files

wzieba committed:
Don't copy gradle.properties-example

It's no longer necessary: default values will be now taken from `defaults.properties` file

wzieba committed:
Set details required for signing configs with data from secrets

wzieba committed:
Remove `android.buildTypes.release.signingConfig = android.signingConfigs.release`

This doesn't seem to be necessary

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Fix declaring signing secrets

wzieba committed:
Merge branch 'trunk' into introduce_secrets_properties

wzieba committed:
Revert "Remove `android.buildTypes.release.signingConfig = android.signingConfigs.release`"

This reverts commit 7551c375a8e70eedf0e7a27fe6e19afb5df2c30f.

wzieba committed:
Keep secrets outside of repository

This is safer: we don't risk accidental include of secrets file when commiting from the repository.

wzieba committed:
Simplify loading secrets

Actually, I think there's no reason for merging `defaults` and `secrets` so this PR simplifies the process

ParaskP7 commented on README.md:
```diff
@@ -8,7 +8,6 @@ If you're a developer wanting to contribute, read on.
 1. Make sure you've installed [Android Studio](https://developer.android.com/studio).
 1. Install npm using [Node Version Manager](https://github.com/nvm-sh/nvm)(nvm), as described in step one from the [Block Editor Quickstart guide](https://developer.wordpress.org/block-editor/getting-started/devenv/#quickstart)
 1. `cd WordPress-Android` to enter the working directory.
-1. `cp gradle.properties-example gradle.properties` to set up the sample app credentials file.
```
Suggestion (💡): I am seeing this section a bit outdated/unnecessary, give-or-take the `npm` instruction. Wdyt if we go ahead and remove it, or replace it with just the `npm` instruction? 🤔

wzieba replied:
I'd leave modifying this instruction to another PR: this already introduces multiple changes, I'd prefer to keep this concise.

ParaskP7 replied:
Now that we are at it, maybe we should at least change the `JDK8/11` reference on step 6... 🤷 😊

ParaskP7 commented on libs/fluxc/gradle.properties:
Question (❓): What about this whole [gradle.properties-example](https://github.com/wordpress-mobile/WordPress-Android/blob/ae41f3f0871193e0a5e72e52ef28b5996ba987ee/libs/fluxc/build.gradle#L57) related configuration, can be removed completely, is it not? 🤔 

PS:
1. Same apply to the [gradle.properties-example](https://github.com/wordpress-mobile/WordPress-Android/blob/ae41f3f0871193e0a5e72e52ef28b5996ba987ee/libs/login/build.gradle#L73) configuration on the `login` module.
2. The `login` module still has an [gradle.properties-example](https://github.com/wordpress-mobile/WordPress-Android/blob/ae41f3f0871193e0a5e72e52ef28b5996ba987ee/libs/login/gradle.properties-example) file that (I guess) needs to be renamed and committed as `gradle.properties`.
3. I think both those `fluxc` and `login` related `gradle.properties` files need to be reviewed a bit more. For example, with the current setup, and the `fluxc` related `gradle.properties` file, the `wp.ENABLE_DATABASE_DOWNGRADE` property points to `true`, maybe, by default, it should be `false` instead. Yes, this `wp.ENABLE_DATABASE_DOWNGRADE` property is being used alongside the `BuildConfig.DEBUG` check, but this is not guaranteed, is it? On the other side, the `fluxc.ENABLE_WPAPI` property is not being used alongside a `BuildConfig.DEBUG` check. Same apply to the `login` module case. Another way to think about them, is to move the into the `secrets.properties`, avoid having them as standalone per module, consider them a secret configuration too, just like we do for the E2E tests, not sure. 🤔

wzieba replied:
I'm not exactly sure how we should proceed with this, but I'd leave decission about default fields to the product team and not modify this in this PR.

To remove `gradle.properties` in modules, as they don't provide that much value, I've decided to move properties to `buildConfigField`. This way, the production code is intact. It's not ideal, but for now I'd leave this as it is.

ParaskP7 replied:
Connecting the dots: [this](https://github.com/wordpress-mobile/WordPress-Android/pull/21442#discussion_r1840076850) + [this](https://github.com/wordpress-mobile/WordPress-Android/pull/21442#discussion_r1840079704)

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -199,6 +196,12 @@ android {
+    Map<String, Object> secretProperties = loadSecrets(
+            project.logger,
+            file("${rootDir}/defaults.properties"),
+            file("${System.getProperty("user.home")}/.configure/wordpress-android/secrets/secrets.properties")
```
Praised (❤️): Nicely done, better be safer! 🥇

wzieba replied:
Kudos to @mokagio for this suggestion! 🙌

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -112,8 +112,6 @@ android {
         testInstrumentationRunner 'org.wordpress.android.WordPressTestRunner'
 
         buildConfigField "boolean", "OFFER_GUTENBERG", "true"
-        buildConfigField "long", "REMOTE_CONFIG_FETCH_INTERVAL", "10"
```
Minor (🔍): I guess this should be removed from the `vanilla` product flavor too, right?

wzieba replied:
Good point, thanks! Addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/21442/commits/d6b817b2f18eaa5777fb5120091e69ebbdc284ac

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -267,9 +270,8 @@ android {
         }
 
         // Also dynamically add additional `buildConfigFields` to our app flavors from any `wp.`/`jp.`-prefixed property in `gradle.properties`
```
Minor (🔍): Consider update the comment to point to `default.properties` instead of `gradle.properties`.

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -287,6 +289,29 @@ android {
 
+    signingConfigs {
+        if (["uploadStoreFile", "uploadStorePassword", "uploadKeyAlias", "uploadKeyPassword"].count { !secretProperties.containsKey(it) } == 0) {
+            logger.info("App signing properties found in secrets.properties, configuring signing for release builds.")
```
Praise (❤️): Awesome you added the logs here (and below)! 💯

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -287,6 +289,29 @@ android {
+                keyAlias = secretProperties.get("uploadKeyAlias")
+                keyPassword = secretProperties.get("uploadKeyPassword")
+            }
+            android.buildTypes.release.signingConfig = android.signingConfigs.release
```
Question (❓): Can you explain what this actually ended-up being necessary for? 🤔

wzieba replied:
Without this line, `release` build type has no `signingConfig` applied. You can verify this by commenting this line and running `signingReport` task

ParaskP7 replied:
Great, thanks for the clarification on that @wzieba ! 🙇

ParaskP7 reviewed (commented):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything (almost) works as expected, good job! 🌟 

-----

I have left few questions (❓), a suggestion (💡) and some minor (🔍) comments for you to consider. I would just like for us to at least discuss the `gradle.properties` and `gradle.properties-example` files on the `login` + `fluxc` modules before merging this. Other than that everything LGTM! 🙏 ✅

wzieba committed:
Remove vanilla-specific build config fields

wzieba committed:
Point to `secrets.properties` instead of `gradle.properties` in code comments

wzieba committed:
Remove FluxC `gradle.properties`

Move build config fields it contained directly to `build.gradle`

wzieba committed:
Remove Login `gradle.properties`

Move build config fields it contained directly to `build.gradle`

wzieba committed:
Move loading secrets to root project extension

This way, secrets are available for the whole project, not only WordPress module

wzieba committed:
Use secret properties to configure `measureBuilds`

Previously, `findProperty` worked out of the box as properties were loaded from `gradle.properties`

wzieba committed:
Remove unused `APP_PN_KEY` build config field

wzieba commented:
Thanks for the review @ParaskP7 ! I applied your suggestions. During the work, I also realized that `measureBuilds` stopped working, so I made some improvements in last commits (654b6b4, f21f8d9) to make it work again. I've also added another test case in PR description.

ParaskP7 commented:
> Thanks for the review @ParaskP7 ! I applied your suggestions. During the work, I also realized that measureBuilds stopped working, so I made some improvements in last commits (https://github.com/wordpress-mobile/WordPress-Android/commit/654b6b461539084675870814e075bc63351da14d, https://github.com/wordpress-mobile/WordPress-Android/commit/f21f8d9c4f5fee404b389885ed70cd4e560581cc) to make it work again. I've also added another test case in PR description.

Awesome, thanks @wzieba , I am currently on it, taking another look at and testing your changes! 🥇

ParaskP7 commented on build.gradle:
```diff
@@ -27,16 +27,39 @@ ext {
     compileSdkVersion = 34
     targetSdkVersion = 34
+
+    secretProperties = loadSecrets(
```
Praise (❤️): Great move! 😄

ParaskP7 commented on build.gradle:
```diff
@@ -27,16 +27,39 @@ ext {
+    }
+}
+
+static def loadPropertiesFromFile(inputFile) {
```
Minor (🔍): Consider adding the (missing) `File` type on `inputFile`, just like we do on `loadSecrets(...)` above, to make the signature of this def more strongly typed.

ParaskP7 commented on libs/fluxc/build.gradle:
```diff
@@ -24,6 +24,9 @@ android {
         consumerProguardFiles 'consumer-rules.pro'
+
+        buildConfigField "boolean", "ENABLE_WPAPI", "false"
+        buildConfigField "boolean", "WP_ENABLE_DATABASE_DOWNGRADE", "true"
```
Warning (⚠️): Let's default that to `false` instead of `true`. I understand what you said [here](https://github.com/wordpress-mobile/WordPress-Android/pull/21442/#discussion_r1838485228) about leaving the decision about default fields to the product team, but IMHO this is atm our problem, us folding `fluxc` into this project and thus us making sure that this default database downgrade will not cause any trouble for the product teams.

What I am trying to say is that having that `true` as default, might cause database problems that during development time (`debug` builds) will go unnoticed and start failing only on release builds (when that's too late), just because the database downgrade wouldn't be enabled by default for release builds.

Note, this, along with the fact that product team will only have that enabled if they copy-paste this specific file from `gradle.properties-example` to `gradle.properties`, makes it seems less of a problem, that is, comparing to now having that enabled by default no matter, and for all.

Does that all makes sense? 🙏

wzieba replied:
I understand the reasoning, but I'd prefer to not change the logic of the runtime app in this PR. If you think this should be _changed_, I'd vote to prepare another PR/issue and discuss it with the team. In this PR, I'd focus only on migration.

wzieba replied:
> Note, this, along with the fact that product team will only have that enabled if they copy-paste this specific file from gradle.properties-example to gradle.properties, makes it seems less of a problem, that is, comparing to now having that enabled by default no matter, and for all.

In the past, product team had this problem as well, as this property was loaded from `gradle.properties-example` if `gradle.properties` file was not found: https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/5dfbbc993d2472f6aad2b9f357021506ecfe7152/fluxc/build.gradle#L52

ParaskP7 replied:
> ...but I'd prefer to not change the logic of the runtime app in this PR.

But, my point is exactly that, with this change we are actually changing the logic of the runtime app in this PR, is it not, am I am missing something @wzieba ? 🤔 

Put it another way:
- **Previously**: If you don't copy-paste `gradle.properties-example` into `gradle.properties` on the `fluxc` module, then you wouldn't get the `WP_ENABLE_DATABASE_DOWNGRADE` as `true`, instead it will default to `false`. I am not expecting anyone cloning/working this app to go and copy-paste `gradle.properties-example` into `gradle.properties` on the `fluxc` module level, unless they need to for testing purposes, nor we do such a thing on CI.
- **Now**: Having the `WP_ENABLE_DATABASE_DOWNGRADE` as `true` on the `build.gradle` level, it will default to `true`, both locally and on CI. When this PR will get merge, every developer will start having a database that has downgrade enabled by default.

ParaskP7 replied:
> In the past, product team had this problem as well, as this property was loaded from gradle.properties-example if gradle.properties file was not found: https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/5dfbbc993d2472f6aad2b9f357021506ecfe7152/fluxc/build.gradle#L52

Oh, interesting, I missed this part, that the build script will actually default to `gradle.properties-example` if `gradle.properties` is missing and load those values instead, interesting, then my comment above is wrong and you're right... 🤔

ParaskP7 commented on libs/login/build.gradle:
```diff
@@ -14,6 +14,12 @@ android {
+        buildConfigField "String", "DEBUG_WPCOM_LOGIN_EMAIL", "\"\""
+        buildConfigField "String", "DEBUG_WPCOM_LOGIN_USERNAME", "\"\""
+        buildConfigField "String", "DEBUG_WPCOM_LOGIN_PASSWORD", "\"\""
+        buildConfigField "String", "DEBUG_WPCOM_WEBSITE_URL", "\"\""
```
Question (❓): I wonder if that's the best thing we can do, I understand it it the fastest, yes. What I am trying to say is that, as you mentioned [here](https://github.com/wordpress-mobile/WordPress-Android/pull/21442/#discussion_r1838485228), this is not ideal, not the same functionality indeed, someone would have to have the Git change visible when needing to update this value, risking committing it by mistake, but I guess, let's deal with that problem when it comes to it, right? 🤷

wzieba replied:
After thinking about this again, actually this might be useful to keep the behaviour of having these properties easy to apply on local machines without risk of commiting. See: 0d1dedb.

And demo that this works fine:

https://github.com/user-attachments/assets/fb77d899-e9c8-47ab-84e7-2925a0f9db6f

ParaskP7 replied:
> After thinking about this again, actually this might be useful to keep the behaviour of having these properties easy to apply on local machines without risk of commiting. See: https://github.com/wordpress-mobile/WordPress-Android/commit/0d1dedb760736fcdd0f5276e51532d9cf81ed2a8.

Awesome, works like a charm and I like the `developers.properties` name you chose too! 💯 

I think this is indeed the best outcome for us and product developers working with that. 👍 

Btw, why not doing the same with `fluxc` too? 🤔

> And demo that this works fine:

Yeap, thanks for the demo, I've been actually using them too, especially after having `login` vendored in, saves time! 💯

wzieba replied:
> Btw, why not doing the same with fluxc too? 🤔

Sure, addressed in f366ab563c0a72d376fd340c143ed73bdef35f3e

ParaskP7 replied:
Awesome, thanks for that @wzieba , everything LGTM! 🚀

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -266,10 +262,9 @@ android {
-        addBuildConfigFieldsFromPrefixedProperties(jetpack, properties, ['wp', 'jp']) // Inherit same properties used from WP first then overwrite with JP-specific ones
+        // Also dynamically add additional `buildConfigFields` to our app flavors from any `wp.`/`jp.`-prefixed property in `secrets.properties`
+        addBuildConfigFieldsFromPrefixedProperties(wordpress, rootProject.secretProperties, ['wp'])
+        addBuildConfigFieldsFromPrefixedProperties(jetpack, rootProject.secretProperties, ['wp', 'jp']) // Inherit same properties used from WP first then overwrite with JP-specific ones
```
Suggestion (💡): Just like you did here, changing `secretProperties` to `rootProject.secretProperties`, I suggest doing the same everywhere (see `singingConfigs { ... }` and `measureBuilds { ... }` blocks.

wzieba replied:
Thanks for spotting this! Actually, I think I'd prefer plain `secretProperties` instead of `rootProject.secretProperties` - the code navigation doesn't work in either case, so anyone looking for source will have to search project either way. And without `rootProject` prefix, it's all shorter and more readable.

ParaskP7 replied:
This makes sense too, since both are working fine, as long as we don't have a mixture of `secretProperties` and `rootProject.secretProperties` it is good with me! 👍

ParaskP7 reviewed (commented):
👋 @wzieba and thanks for the changes, I went for another look, left some more comments for you, but mostly, I would like us to discuss this one warning (⚠️), that is, before merging this to `trunk`. 🙏

wzieba committed:
Add type in `loadPropertiesFromFile` signature

wzieba committed:
Load debug developer properties from local file

Brings back behaviour, previously supported by copying gradle.properties-example. In this case, a new file is used (developer.properties) to make sure it's separated from build configuration

wzieba committed:
Remove unnecessary `rootProject` prefix

wzieba commented on libs/login/build.gradle:
```diff
@@ -26,6 +35,22 @@ android {
     }
 }
 
+static def loadDeveloperProperties(File defaultsFile, File developerFile) {
```
These methods are similar to how we handle secrets, but sharing methods between Gradle projects is not really easy, so to reduce complexity and because these methods are small I've decided to write them again

wzieba committed:
FluxC: Load debug developer properties from local file

Brings back behaviour, previously supported by copying gradle.properties-example. In this case, a new file is used (developer.properties) to make sure it's separated from build configuration

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-Android/milestone/298">25.8</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio commented on .configure:
```diff
@@ -1,11 +1,11 @@
-      "file": "android/WPAndroid/gradle.properties",
-      "destination": "gradle.properties",
+      "file": "android/WPAndroid/secrets.properties",
+      "destination": "~/.configure/wordpress-android/secrets/secrets.properties",
```
💙
